src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -782,6 +924,39 @@ private void AddMethodReference(MethodDesc method)
             _dependencies.Add(_compilation.NodeFactory.MethodEntrypoint(method));
         }
 
+        private void AddTypeReference(TypeDesc type, bool constructed)
+        {
+            // WebAssemblyImporter will rather arbitrarily try to generate types as constructed.
+            // Stomp over the choice and only allow this if it remotely makes sense.
+            constructed = constructed & ConstructedEETypeNode.CreationAllowed(type);
+
+            AddTypeDependency(type, constructed);
+
+            if (!type.IsGenericDefinition)
```

Comment:
I wouldn't expect needing to manually mark fields here. I'd expect that to come from other usage or the dependency node infrastructure.

Reply:
This code was just ported over from the cpp codegen, are you sure this gets taken care of elsewhere?

Reply:
CppCodegen is only partially hooked up into the dependency analysis infrastructure because the code pre-dates the dependency analysis. There are hacks in place that try to shoehorn the dependency analysis model into the model of CppCodegen code emission. The entire `AddTypeReference` method is one of such CppCodegen hacks.

The way CppCodegen is structured, we need an EEType for all fields of a type. Other codegen backends we have (RyuJIT, UTC) don't have such requirements. The runtime definitely doesn't need them. Fields get dissolved into (implicit) knowledge of their storage size and their GC layout description. Nothing else needs to survive the compilation.

Reply:
It might be better to reference the [IL scanner](https://github.com/dotnet/corert/blob/5d7e4af8df1262b7b8ab8cd0b805c9f498d3493a/src/ILCompiler.Compiler/src/IL/ILImporter.Scanner.cs) code to get a sense of actual dependencies needed. The purpose of the IL scanner is to predict what RyuJIT compilation will actually need. It does pretty much everything a codegen backend would do (dependency-wise), minus actually generating code.